### PR TITLE
Fix generate with `inputs_embeds` as input

### LINKED ISSUE #32479: Gemma2ForCausalLM: ValueError in prepare_inputs_for_generation when using custom input embeddings
### System Info

- `transformers` version: 4.44.0
- Platform: Linux-5.4.0-189-generic-x86_64-with-glibc2.31
- Python version: 3.10.14
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.32.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.3.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: NO
- Using GPU in script?: YES
- GPU type: NVIDIA RTX A6000

### Who can help?

@ArthurZucker
@gante 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

model_name="google/gemma-2-2b-it"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    device_map="auto",
    torch_dtype=torch.bfloat16,
)

messages = [
    {"role": "user", "content": "Write me a poem about Machine Learning."},
]
template = tokenizer.apply_chat_template(messages, tokenize=False)
input_ids = tokenizer(template, return_tensors="pt").to("cuda")

embedding_layer = model.get_input_embeddings()
inputs_embeds = embedding_layer(input_ids["input_ids"])

outputs = model.generate(
    inputs_embeds=inputs_embeds,
    max_new_tokens=32,
)
```

### Expected behavior

I'm encountering an error when attempting to use custom input embeddings with the Gemma2ForCausalLM model's `.generate()` method. Specifically, I'm modifying the input embeddings for the model and then trying to generate output from these custom embeddings.

**Expected behavior:**
When using the `inputs_embeds` argument to supply custom input embeddings, the `.generate()` function should be able to process these embeddings and produce a generated output. This is what happens in other models such as "meta-llama/Meta-Llama-3-8B-Instruct".

**Actual behavior:**
The `.generate()` method raises a `ValueError` in the `prepare_inputs_for_generation` function. The error suggests that the `inputs_embeds` tensor doesn't have the expected shape, resulting in a "too many values to unpack (expected 2)" error.

The error trace looks like the following:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[5], line 21
     18 embedding_layer = model.get_input_embeddings()
     19 inputs_embeds = embedding_layer(input_ids["input_ids"])
---> 21 outputs = model.generate(
     22     inputs_embeds=inputs_embeds,
     23     max_new_tokens=32,
     24 )

File <full_path>/miniforge3/envs/myenv/lib/python3.10/site-packages/torch/utils/_contextlib.py:115, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    112 @functools.wraps(func)
    113 def decorate_context(*args, **kwargs):
    114     with ctx_factory():
--> 115         return func(*args, **kwargs)

File <full_path>/miniforge3/envs/myenv/lib/python3.10/site-packages/transformers/generation/utils.py:2024, in GenerationMixin.generate(self, inputs, generation_config, logits_processor, stopping_criteria, prefix_allowed_tokens_fn, synced_gpus, assistant_model, streamer, negative_prompt_ids, negative_prompt_attention_mask, **kwargs)
   2016     input_ids, model_kwargs = self._expand_inputs_for_generation(
   2017         input_ids=input_ids,
   2018         expand_size=generation_config.num_return_sequences,
   2019         is_encoder_decoder=self.config.is_encoder_decoder,
   2020         **model_kwargs,
   2021     )
   2023     # 13. run sample (it degenerates to greedy search when `generation_config.do_sample=False`)
-> 2024     result = self._sample(
   2025         input_ids,
   2026         logits_processor=prepared_logits_processor,
   2027         logits_warper=prepared_logits_warper,
   2028         stopping_criteria=prepared_stopping_criteria,
   2029         generation_config=generation_config,
   2030         synced_gpus=synced_gpus,
   2031         streamer=streamer,
   2032         **model_kwargs,
   2033     )
   2035 elif generation_mode in (GenerationMode.BEAM_SAMPLE, GenerationMode.BEAM_SEARCH):
   2036     # 11. prepare logits warper
   2037     prepared_logits_warper = (
   2038         self._get_logits_warper(generation_config, device=input_ids.device)
   2039         if generation_config.do_sample
   2040         else None
   2041     )

File <full_path>/miniforge3/envs/myenv/lib/python3.10/site-packages/transformers/generation/utils.py:2975, in GenerationMixin._sample(self, input_ids, logits_processor, stopping_criteria, generation_config, synced_gpus, streamer, logits_warper, **model_kwargs)
   2969 model_kwargs = self._get_initial_cache_position(input_ids, model_kwargs)
   2971 while self._has_unfinished_sequences(
   2972     this_peer_finished, synced_gpus, device=input_ids.device, cur_len=cur_len, max_length=max_length
   2973 ):
   2974     # prepare model inputs
-> 2975     model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
   2977     # prepare variable output controls (note: some models won't accept all output controls)
   2978     model_inputs.update({"output_attentions": output_attentions} if output_attentions else {})

File <full_path>/miniforge3/envs/myenv/lib/python3.10/site-packages/transformers/models/gemma2/modeling_gemma2.py:1223, in Gemma2ForCausalLM.prepare_inputs_for_generation(self, input_ids, past_key_values, attention_mask, inputs_embeds, cache_position, position_ids, use_cache, **kwargs)
   1221 if isinstance(past_key_values, HybridCache) and attention_mask.ndim == 2:
   1222     if inputs_embeds is not None:
-> 1223         batch_size, sequence_length = inputs_embeds.shape
   1224         device = inputs_embeds.device
   1225     else:

ValueError: too many values to unpack (expected 2)
```

### PR DESCRIPTION
**Pull request by molbap**
# What does this PR do?

Seems recent refactor broke generation based on `inputs_embeds` if I understand correctly.
This aims at catching correctly the `sequence_length` by using the passed `model_inputs`, creating the correctly-sized `causal_mask` in generation step. 

Core change is 

```python
        if inputs_embeds is not None and cache_position[0] == 0:
            model_inputs = {"inputs_embeds": inputs_embeds, "input_ids": None}
        else:
            # The clone here is for the same reason as for `position_ids`.
            model_inputs = {"input_ids": input_ids.clone(memory_format=torch.contiguous_format), "inputs_embeds": None}
        # or any Cache class
        if isinstance(past_key_values, HybridCache) and attention_mask.ndim == 2:
            if model_inputs["inputs_embeds"] is not None:
                batch_size, sequence_length, _ = model_inputs["inputs_embeds"].shape
                device = model_inputs["inputs_embeds"].device
            else:
                batch_size, sequence_length = model_inputs["input_ids"].shape
                device = model_inputs["input_ids"].device
```

Fixes https://github.com/huggingface/transformers/issues/32479

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Thanks, will be included in a patch. 
Can you add a test in the generate tests this should clearly have been caught

**Comment by molbap** [comment]
Yes doing that, a bit surprised that this slipped through

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -1265,17 +1265,18 @@ def prepare_inputs_for_generation(
...
         else:
-            model_inputs = {"input_ids": input_ids}
+            # The clone here is for the same reason as for `position_ids`.
+            model_inputs = {"input_ids": input_ids.clone(memory_format=torch.contiguous_format), "inputs_embeds": None}
 
         if isinstance(past_key_values, StaticCache) and attention_mask.ndim == 2:
-            if inputs_embeds is not None:
-                batch_size, sequence_length = inputs_embeds.shape
-                device = inputs_embeds.device
+            if model_inputs["inputs_embeds"] is not None:
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1274]: I don't understand why we don't use `input_embeds` ?

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -1265,17 +1265,18 @@ def prepare_inputs_for_generation(
...
+                batch_size, sequence_length, _ = model_inputs["inputs_embeds"].shape
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1275]: same here

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -1265,17 +1265,18 @@ def prepare_inputs_for_generation(
...
+                device = model_inputs["inputs_embeds"].device
             else:
-                batch_size, sequence_length = input_ids.shape
-                device = input_ids.device
+                batch_size, sequence_length = model_inputs["input_ids"].shape
+                device = model_inputs["input_ids"].device
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1279]: and here as well

File: tests/test_modeling_common.py [hunk 4]
```diff
@@ -2819,6 +2819,53 @@ def test_inputs_embeds_matches_input_ids(self):
...
+            pad_token_id = config.pad_token_id if config.pad_token_id is not None else 1
+
+            wte = model.get_input_embeddings()
+            if not self.is_encoder_decoder:
+                input_ids = inputs["input_ids"]
+                # some models infer position ids/attn mask differently when input ids
+                # by check if pad_token let's make sure no padding is in input ids
+                not_pad_token_id = pad_token_id + 1 if max(0, pad_token_id - 1) == 0 else pad_token_id - 1
+                input_ids[input_ids == pad_token_id] = not_pad_token_id
+                del inputs["input_ids"]
```
Comment (ArthurZucker) [inline tests/test_modeling_common.py:2845]: let's just pass our own torch.ones() should be easier

File: tests/test_modeling_common.py [hunk 5]
```diff
@@ -2819,6 +2819,53 @@ def test_inputs_embeds_matches_input_ids(self):
...
+                inputs_embeds = wte(input_ids)
+                out_ids = model.generate(input_ids=input_ids, **inputs, max_new_tokens=2)[:, -2:]
+                out_embeds = model.generate(inputs_embeds=inputs_embeds, **inputs, max_new_tokens=2)
+            else:
+                encoder_input_ids = inputs["input_ids"]
+                decoder_input_ids = inputs.get("decoder_input_ids", encoder_input_ids)
+                encoder_input_ids[encoder_input_ids == pad_token_id] = max(0, pad_token_id + 1)
+                decoder_input_ids[decoder_input_ids == pad_token_id] = max(0, pad_token_id + 1)
+                del inputs["input_ids"]
+                inputs.pop("decoder_input_ids", None)
```
Comment (ArthurZucker) [inline tests/test_modeling_common.py:2855]: same here not sure we want to mix padding as well

Comment (molbap) [reply tests/test_modeling_common.py:2855]: ah, found it weird too - copied the behaviour from another common generate test, just above in the file, figured not to change too many things

File: src/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (molbap) [reply src/transformers/models/llama/modeling_llama.py:1274]: It's the same here: we used `inputs_embeds` to initialize the `model_inputs` dict and we never reuse it
The issue is that if a user generates with input_embeds, they persist, and `sequence_length` is taken as dimension -2 of inputs_embeds for the causal mask, which will cause a shape mismatch

File: tests/test_modeling_common.py (hunk 5 above)
Comment (molbap) [reply tests/test_modeling_common.py:2855]: from `test_inputs_embeds_matches_input_ids` to be specific (which tests forward, not generate)

Comment (ArthurZucker) [reply tests/test_modeling_common.py:2855]: ah ok

File: src/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:1274]: mmm yeah but it's a bit weird generate never gives both inputds or input embeds. It cool that we explicit it, we should just end up with 1 if else!

Comment (molbap) [reply src/transformers/models/llama/modeling_llama.py:1274]: Well it seems I ended up with inputs_embeds and a size-0 input_ids, but not None :thinking: so wanted to warrant this. For the sole if/else ah yes you're right, you mean like

```python
if isinstance(past_key_values, StaticCache) and attention_mask.ndim == 2:
    if inputs_embeds is not None and cache_position[0] == 0:
        model_inputs = {"inputs_embeds": inputs_embeds, "input_ids": None}
        batch_size, sequence_length, _ = model_inputs["inputs_embeds"].shape
        device = model_inputs["inputs_embeds"].device
    else:
        model_inputs = {"input_ids": input_ids.clone(memory_format=torch.contiguous_format), "inputs_embeds": None}
        batch_size, sequence_length = model_inputs["input_ids"].shape
        device = model_inputs["input_ids"].device
```

will update!

Comment (molbap) [reply src/transformers/models/llama/modeling_llama.py:1274]: ah no, that kills the model_inputs initialization - you had smth in mind for the only if/else?

Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:1274]: not sure no, just creating the model_inputs after the if else using input_embeds and input_id, not uregnet le'ts just merge this!

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (molbap) [reply src/transformers/models/llama/modeling_llama.py:1274]: alright! yes, we can do it later - merging, thanks!

**Comment by SunMarc** [comment]
Hey @molbap, I think that we forgot to skip a few tests related to this PR : https://app.circleci.com/pipelines/github/huggingface/transformers/100558/workflows/0c640751-f999-4b6a-9f8d-ebc7f869c50f/jobs/1339096?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary . Can you confirm me that we indeed need to skip them just like bert model ? I'll open a PR if that's the case ! Not sure why these tests are only triggered on my PR: https://github.com/huggingface/transformers/pull/32627

**Comment by molbap** [comment]
Hi @SunMarc , there was an issue with the fetching of tests and it is causing a lot of failures currently, opening a PR to fix it - in the meantime you can indeed skip them!

**Comment by molbap** [comment]
In the end #32651 was also needed to fix breaking tests introduced here, FYI!

**Pull request merged**